### ISSUE TITLE
seth: use ethers.js instead of ethabi

### ISSUE THREAD

**Issue by MrChico** [dapphub/dapptools#298]
This would also bring support for ABIEncoder v2, which ethabi doesn't seem to care much for

**Comment by MrChico** [comment]
ethabi is horrendously slow!
```
FUNCABI=$(seth --abi-function-json 'sixteenArgs(address,address,uint256,uint256,uint256,uint256,address,uint8,bytes32,bytes32,address,address,uint256,uint256,bool,uint8)')
time ethabi encode function <(echo $FUNCABI) sixteenArgs -p 2C33B04ae6B405dEE94a6503dbe6a7CaD3f082Ca -p 2C33B04ae6B405dEE94a6503dbe6a7CaD3f082Ca -p 0000000000000000000000000000000000000000000000001bc16d674ec80000 -p 00000000000000000000000000000000000000000000000001671455cef6f770 -p 0000000000000000000000000000000000000000000000000000000000000000 -p 0000000000000000000000000000000000000000000000000000000000000000 -p c3455912Cf4bF115835A655c70bCeFC9cF4568eB -p 28 -p 7b93f2490c1833e65f9bfa58cd50e9f097edc408e65ebe88a5551dfb53ec3755 -p 4ab85d434db4881066e773411c6cf6b86966f4b5860f475b4dba2fa5608302ce -p 2C33B04ae6B405dEE94a6503dbe6a7CaD3f082Ca -p 2ef865ca017912c872826e25736a7a11f94235af -p 0000000000000000000000000000000000000000000000000000000000000000 -p 0000000000000000000000000000000000000000000000000000000000000000 -p 1 -p 28  -l
28.04s user 2.46s system 99% cpu 30.523 total
```

**Comment by MrChico** [comment]
I guess rust can have memory leaks too? https://github.com/paritytech/ethabi/issues/161

**Comment by xwvvvvwx** [comment]
Rust lets you make reference cycles: https://doc.rust-lang.org/stable/book/ch15-06-reference-cycles.html

**Comment by asymmetric** [comment]
@MrChico what's `EVM.ABI`?

**Comment by xwvvvvwx** [comment]
https://github.com/dapphub/dapptools/blob/master/src/hevm/src/EVM/ABI.hs ?

**Comment by asymmetric** [comment]
We would have to make that available as a binary then.

**Comment by livnev** [comment]
Need to add this functionality to `hevm-cli`.

**Comment by livnev** [comment]
> We would have to make that available as a binary then. How portable are Haskell binaries? Can we distribute them as statically linked or would people need a whole Haskell distribution?

What is wrong with the current nix way we are distributing `dapptools`?

**Comment by asymmetric** [comment]
IIUC what's being proposed here is to use a Haskell binary instead of a Rust binary. Since `ethabi` is used in `seth`, we'd need to think what adding a Haskell dependency means in terms of packaging `seth`, mainly size-wise.

Currently not everyone will install `hevm`, if we use `EVM.ABI` everyone will have to install (parts of) it.

Or am I completely off the mark? 

But to answer your question @livnev:

> What is wrong with the current nix way we are distributing dapptools?

Nothing! :)

**Comment by asymmetric** [comment]
Also, we'd have to expose `EVM.ABI` as a binary I suppose?

Not against this proposed change, just trying to explicitate what's involved :)

**Comment by livnev** [comment]
> IIUC what's being proposed here is to use a Haskell binary instead of a Rust binary. Since `ethabi` is used in `seth`, we'd need to think what adding a Haskell dependency means in terms of packaging `seth`, mainly size-wise.
> 
> Currently not everyone will install `hevm`, if we use `EVM.ABI` everyone will have to install (parts of) it.
> 
> Or am I completely off the mark?

No this is exactly right, I remember a conversation once before where there was hesitation about having to distribute `hevm` as a new dependency of `seth`.

**Comment by livnev** [comment]
> Also, we'd have to expose `EVM.ABI` as a binary I suppose?
> 
> Not against this proposed change, just trying to explicitate what's involved :)

We would add a cli for the required features to `hevm-cli` I think.

**Comment by MrChico** [comment]
Right... I didn't think of that we don't include hevm in `seth` right now. I'm open to exploring more light weight alternatives, if there's concern for including `hevm` as a dependency.

**Comment by asymmetric** [comment]
> I'm open to exploring more light weight alternatives, if there's concern for including hevm as a dependency.

Maybe we should start thinking about maintaining our own fork, since ethabi might stop being maintained by Parity.

**Comment by MrChico** [comment]
It seems `ethers.js` does everything we want, which we already depend on anyway. I'm reimplementing `seth-calldata`, `seth---abi-decode` and `seth---event-decode` with this instead

**Comment by cecton** [comment]
Heya o/ Parity here. I have investigated the issue and we can fix the issue rather quickly now. But ethabi is not going to be maintained by us for long as it will be moved to OpenEthereum at some point https://www.parity.io/parity-ethereum-openethereum-dao/

Let me know if you prefer to move to ethers.js or if you want me to fix ethabi so I can prioritize

**Comment by MrChico** [comment]
Heya @cecton thanks for chiming in. I think it's a good for us to switch to ethers.js. 
ethabi has served us well, but with its uncertain future I think it's better for us to rely on ethers.js. (which gives us one less dependency)

**Comment by cecton** [comment]
Ah damn, I saw it too late :confounded:  Ok, I hope we will be able to help in the future

**Comment by asymmetric** [comment]
@cecton If the fix is not too onerous for you, I think we'd still like to see it land, as that leaves us with the option open to switch back to ethabi in the future.

**Comment by cecton** [comment]
@asymmetric it's done already :smiley:  I wait for someone to review. Glad I could help

**Comment by cecton** [comment]
The changes (fix) have been released: v10.0.0 @asymmetric @MrChico

**Comment by dbrock** [comment]
For the record I think it makes a lot of sense to bundle Hevm with Seth and in fact I think moving more and more towards depending on Hevm whereever it makes sense seems very logical. It's much nicer to depend on something we control and maintain than a garbage collection of random third party utilities, hehe.
